### PR TITLE
app.go cleanup

### PR DESCRIPTION
**Pull request by Anmol1696**
PR for #1557 

Cleanup to make `app.go` easier and more similar to cleaner upgrades and keeper
* `upgrades/` directory to hold various chain upgrade functions
* `keepers/` dir to hold all keeper creation, just moved keeper defination from `app.go` to keeper and another struct to be passed to upgrades.

Note: Core logic for is just moved to new structure, not much refactor.

### CODE REVIEW

**Comment by okwme** [comment]
Thank you @Anmol1696 !
appreciate the new formatting : )
will need to check on status of new features being added to main to confirm whether to merge this before or after (global fee + bypass min fee specifically)
please standby!

**Comment by yaruwangway** [comment]
probably after global fee and ica auth module integrated...

**Comment by Anmol1696** [comment]
Cool no problem, will wait till then before resolving all following conflicts.

Let me know if there is anything else to be added to this PR.

**Comment by Anmol1696** [comment]
I guess still waiting on some of the PRs. These are the PRs i should be waiting for right?
* https://github.com/cosmos/gaia/pull/1488
* https://github.com/cosmos/gaia/pull/1530

**Comment by yaruwangway** [comment]
> I guess still waiting on some of the PRs. These are the PRs i should be waiting for right?
> 
> * [feat: global fee module #1488](https://github.com/cosmos/gaia/pull/1488)
> * [feat: integrating ica to rho #1530](https://github.com/cosmos/gaia/pull/1530)
yes, @Anmol1696 , thank you for the pr !  it still takes a while to merge the feat prs. we need the dependencies bump to new sdk before continue. i will @ you at this PR when the other two prs are merged.

**Comment by faddat** [comment]
I like this PR.

I'd like to take some time tomorrow, and compare it to Osmosis' app.go, I really like how they've structured things.

**Review by faddat — APPROVED** [review]
I like this style, and have been working with it at Osmosis. 

I ran through the PR just now and overall, I just think that it'll benefit us re: usability and readability.

I am also sure that it's going to need a little bit of restructuring because of the upgrade 46. Would you like to do it @Anmol1696 or should I pull your branch and make the changes in a new branch?

**Comment by Anmol1696** [comment]
@faddat i can take up the restructure, if you can point me towards it, i want to take it up.

**Comment by faddat** [comment]
So here are two examples:

https://github.com/notional-labs/craft

https://github.com/osmosis-labs/osmosis

They are both very similar, and I think it is a great way to reduce the complexity of app.go

File: app/keepers/keepers.go
Comment (puneet2019) [inline app/keepers/keepers.go]: This is duplicate of line: https://github.com/cosmos/gaia/pull/1580/files#diff-1ac782d2fab1d48ad7f7ef06cf2d58b37c99b6b4c98879a5ad86a985b73fad58R244-R251

Fixed in main branch in this PR:
https://github.com/cosmos/gaia/pull/1725

**Comment by yaruwangway** [comment]
@Anmol1696 , please solve the conflicts and we can review and merge the pr ! thank you! 👍

**Comment by Anmol1696** [comment]
Fixed the conflicts, some of the tests seem to be failing, but that has nothing to do with the PR though

**Comment by Anmol1696** [comment]
@yaruwangway do you think the CI errors is something that is fixed somewhere else?
If this is the first time this is happening can have a deeper look at it.

File: app/keepers/keepers.go
Comment (puneet2019) [inline app/keepers/keepers.go:372]: @okwme are there plans to use `appKeepers.RouterModule` ?

**Comment by yaruwangway** [comment]
> @yaruwangway do you think the CI errors is something that is fixed somewhere else? If this is the first time this is happening can have a deeper look at it.

Hi, i can have a look

**Comment by yaruwangway** [comment]
@Anmol1696 
i think its here in `app/keepers/keeper.go`:
`appKeepers.ICAModule = ica.NewAppModule(nil, &appKeepers.ICAHostKeeper)` ->
`appKeepers.ICAModule = ica.NewAppModule(&app.ICAControllerKeeper, &appKeepers.ICAHostKeeper)`

**Comment by Anmol1696** [comment]
Ahh thanks @yaruwangway sorry my bad, should have been more thorough, i was able to replicate the issue locally, just not enough time to look into it. 
Thanks will fix and push

File: app/keepers/keepers.go
Comment (puneet2019) [reply app/keepers/keepers.go:372]: @yaruwangway

**Comment by yaruwangway** [comment]
Hi @Anmol1696 , is this pr ready for review ?

**Comment by Anmol1696** [comment]
Yup, somehow i am not able to set it for review? The reviewers list basically disappeared.

File: app/upgrades/v8/constants.go
Comment (yaruwangway) [inline app/upgrades/v8/constants.go:1]: this should be v7 actually
we can create a new v8 file for later v8

File: app/upgrades/v8/upgrades.go
Comment (yaruwangway) [inline app/upgrades/v8/upgrades.go]: this upgrade is for v7, 
 @glnro will take care of v8 upgrade handler later.

**Comment by yaruwangway** [comment]
> Yup, somehow i am not able to set it for review? The reviewers list basically disappeared.

Hi, I set it. no worries. thank you for this big refactor !  👍

File: app/upgrades/v8/constants.go
Comment (Anmol1696) [reply app/upgrades/v8/constants.go:1]: ok will change it, but in the app.go, the constant name was v8-rho, so thought this handler is for v8.

File: app/upgrades/v8/upgrades.go
Comment (Anmol1696) [reply app/upgrades/v8/upgrades.go]: moved this function to v7, and added empty v8 upgrade functions

File: app/upgrades/v8/constants.go
Comment (yaruwangway) [reply app/upgrades/v8/constants.go:1]: yeah, we will add v8 upgrade handler later.  sorry its confusing.

**Comment by Anmol1696** [comment]
The code coverage seems to have droped more than 5%.
If you guys think this PR needs more tests, would be happy to add them and bring it up.

I was hopping to add better testing in another PR though, since it would require more discussions on testing designs for the app itself.

**Comment by yaruwangway** [comment]
feel free to add more tests ! 👍

**Comment by Anmol1696** [comment]
> feel free to add more tests ! 👍

You want it in this PR? I was thinking of having seperate PR for tests.

**Comment by yaruwangway** [comment]
hi @Anmol1696 , maybe can [exclude/ignore](https://docs.codecov.com/docs/ignoring-paths#:~:text=You%20can%20use%20the%20top,will%20be%20skipped%20during%20processing.) some of the files [codecov indicating that this pr not covered](https://app.codecov.io/gh/cosmos/gaia/pull/1580) ? `modules.go` , `types.go` ,  `constants.go`
so we can merge this pr early.

**Comment by Anmol1696** [comment]
The complaint seems to be directly for the `app/app.go` file itself, ignoring other files does not seem to help this case. I would not want to ignore the `app.go` file itself for this though. I did try to ignore the files and it does not seem to help.

This seems to be special case where there is an overhaul of the full `app.go` file, hence the CI for having 5% change does not make sense (for this PR only though).

If we dont want to use admin to merge the PR, we could have back to back PRs, in this PR we ignore the `app.go`, and after it is merged, next PR to remove the ignore.

Let me know how you want me to proceed here @yaruwangway, or if you have any other suggestions.

File: app/app.go
Comment (Pantani) [inline app/app.go:401]:
```suggestion
func (app *GaiaApp) OnTxSucceeded(_ sdk.Context, _, _ string, _ []byte, _ []byte) {}
```

Comment (Pantani) [reply app/app.go:401]: We can also do the same for `OnTxFailed`. WDYT?

```go
func (app *GaiaApp) OnTxFailed(_ sdk.Context, _, _ string, _ []byte, _ []byte) {}
```

**Review by Pantani — CHANGES_REQUESTED** [review]
I think we can remove the [`EmptyAppOptions`](https://github.com/cosmos/gaia/pull/1580/files#diff-0f1d2976054440336a576d47a44a37b80cdf6701dd9113012bce0e3c425819b7L1043-L1049) from the `app.go`. It is duplicated in the `helpers/test_helpers.go`

**Review by yaruwangway — APPROVED** [review]

**Comment by yaruwangway** [comment]
@mmulji-ic can we force merge ?

**Comment by okwme** [comment]
ah sorry @Anmol1696, every PR that gets merged into main presents a new set of conflicts. If you could help get the latest ones resolved I'll merge the PR before the next ones, thank you!

**Comment by Anmol1696** [comment]
Yup let me take it up today and finish this up.
Thanks @yaruwangway for the lint fixes.

File: app/app.go
Comment (Anmol1696) [reply app/app.go:401]: linters are not throwing errors... so i dont see a point of switching this though

Comment (Anmol1696) [reply app/app.go:401]: About removing the `EmptyAppOptions`, makes sense to remove it, but I think it should be part of the next PR, in this PR I just wanted to focus on refactor of the app itself to the new structure, already hard to review the PR.
I can open a small chore PR.

**Comment by yaruwangway** [comment]
Hi @Anmol1696, thank you! Presently, i did not merge [this commit of bumping ibc](https://github.com/cosmos/gaia/commit/850aeb7faf121c033432e701c91b895068a3dc0e) in your PR, you can neglect this commit, we can do this commit again after your PR merged!

**Comment by yaruwangway** [comment]
@Anmol1696 
you can just comment out `TestAccAddress = icatypes.GenerateAddress(sdk.Context{}, ibctesting.FirstConnectionID, TestPortID)
` to fix the err. this is not even used.

**Comment by Anmol1696** [comment]
Hmm somehow the commits got messed up, i tried to hard push a force reset to `2e231b7` on my branch and a commit to comment out the test function in the latest commit, but now when i checkout to `2e231b7`, the build still seems to be broken, although main itself i building.

A little confused here.

**Comment by Anmol1696** [comment]
It seems i can try to do the merge again though and see what happens on my branch, but l am not sure what just happened here to this PR

**Comment by yaruwangway** [comment]
> It seems i can try to do the merge again though and see what happens on my branch, but l am not sure what just happened here to this PR

so you force pushed? I think there are two commits are missing. i can fix it.

**Comment by Anmol1696** [comment]
well i forced push to remove your commits from my branch, but turns out `[2e231b7]` commit itself was broken.. i am just curious how it got there now. Force pushing to remove unwanted commits basically

**Comment by yaruwangway** [comment]
you only removed two commits from mine, there are three.

**Comment by yaruwangway** [comment]
Would you like to fix this PR or you prefer I do ?

**Comment by Anmol1696** [comment]
yup i was asking since that was a merge commit to update my branch with main, i kept it. 
I will remove that commit as well then, but my question was around that only, how was a merge commit wrong.

**Comment by Anmol1696** [comment]
well the fix just requiers a force push back to orginial and then re-merge the main branch. Right?
I will do it

**Comment by yaruwangway** [comment]
ha, if you mean why merge failed the test, 
this is due to [this commit of bumping ibc](https://github.com/cosmos/gaia/commit/850aeb7faf121c033432e701c91b895068a3dc0e) in main.

from present commit, you can further commit to convert back the ibc-go version from v5.0.0 to v5.0.0-rc2. it will fix some of the tests. and we will bump the ibc again after your branch merge.

otherwise, you can reset back to `1cfbbb446905c0a02652aa579af60df9fed4180f` and re-merge main. mainly you need to distribute [this commit of bumping ibc](https://github.com/cosmos/gaia/commit/850aeb7faf121c033432e701c91b895068a3dc0e) into the refactor.

**Comment by Anmol1696** [comment]
Ahh cool cool, my thought was that the merge of main branch was correct, because this bump of ibc was in a later commit of fix linting which i rolled back as well.
Let me just reset and merge main branch completly then. Thanks

**Comment by Anmol1696** [comment]
Done with the branch again. can you guys have a look: @yaruwangway @okwme

File: app/app.go
Comment (Anmol1696) [reply app/app.go:401]: @Pantani does that sound ok?

**Comment by yaruwangway** [comment]
merged! thanks! @Anmol1696

**Pull request merged**
